쥬스 메이커 [STEP 2] safari, Red

Siwon-L opened the pull request:
@stevenkim18
안녕하세요 steven 😁 😃
2_group의 safari, Red 입니다. 잘 부탁드립니다!
쥬스메이커 STEP 2 PR 보냅니다
![](https://i.imgur.com/83NLkZc.png)


---
## 문제상황과 해결방법
- 어려 버튼에 대한 액션을 하나의 IBAction으로 처리해줄 것인가 
- 어떤 방식으로 에러 처리를 해줄 것인가 
- 다른 상황에 보이는 얼럿에 대한 함수를 분리할 것인가? 통합할 것인가?
- 화면전환을 모달을 사용할 것인가 내비게이션을 사용할 것인가? 

### 어려 버튼에 대한 액션을 하나의 IBAction으로 처리해줄 것인가
프로젝트를 구상할 때부터 7개의 버튼마다 IBAction 함수를 만들어주는 것보다 하나의 함수로 처리하는 방식을 생각했습니다. 때문에 버튼마다 차별되는 값을 사용해야 했고, 저는 버튼에 tag를 이용하는 방식을, Red는 UILabel 타입을 상속받아 Juice 타입의 변수를 추가하여 커스텀 라벨을 만들어보는 방법을 제시하였습니다. 하지만, tag를 이용하는 방식은 tag는 정수타입이기 때문에 하드코딩에 우려가 있고, tag의 의도를 파악하지 못한 상태에서 사용하는것은 옳지 못하다 생각했습니다. 또한, UILabel 타입을 상속받아 Juice 타입의 변수를 추가하는 방식은 배우지 못한 이니셜라이져 키워드와 viewDidLoad()에서 값을 넣어주는 행위 등 추가적인 행위가 생겼고, 때문에 버튼의 Title를 이용하여 `@IBOutlet`으로 버튼객체를 선언해준 뒤,
```swift
func matchButtonToJuice(_ button: UIButton) -> Juice? {
        var juice: Juice?
        switch button {
        case strawberryJuiceButton:
            juice = .strawberryJuice
        case bananaJuiceButton:
            juice = .bananaJuice
        case pineappleJuiceButton:
            juice = .pineappleJuice
        case kiwiJuiceButton:
            juice = .kiwiJuice
        case mangoJuiceButton:
            juice = .mangoJuice
        case strawberryAndBananaJuiceButton:
            juice = .strawberryAndBananaJuice
        case mangoAndKiwiJuiceButton:
            juice = .mangoAndKiwiJuice
        default:
            juice = nil
        }
        return juice
    }
```
위와 같이 스위치문법을 이용한 함수를 만들어 IBAction에서 사용하였습니다.

### 어떤 방식으로 에러 처리를 해줄 것인가
takeOrder 함수에서 재고가 부족한 경우에 오류 메세지(오류 얼럿)을 띄워주어야 하는데, step1 에서는 오류의 종류가 다양하지 않고, 해결이 가능한 오류이기 때문에 오류처리 구문을 사용하지 않았습니다. step2로 넘어오면서 viewController 에서 보았을 때, takeOrder 라는 함수에서 오류가 발생할 수 있다는 상황을 적극적으로 표현해 보고 싶어, JuiceMakeError 라는 타입을 설정하고, orderAndCheck 라는 에러 상황을 Bool 값으로 전달해 주는 함수를 구현했습니다. 

### 다른 상황에 보이는 얼럿에 대한 함수를 분리할 것인가? 통합할 것인가?
이번 프로젝트에서 보이는 얼럿을 두 가지로 구현했습니다. 음료가 나왔을 때 알려주는 용도의 얼럿, 재고가 없어 재고 수정에 대한 사용자의 의견을 묻는 얼럿 처음엔 두 얼럿을 하나의 함수로 구현하여 함수 내부에서 조건문을 이용해 하단에 표시되는 버튼만 다르게 해주는 방법을 채택했습니다. 하지만, 에러 처리가 들어가면서 둘로 분리하여, 각각 맞는 상황에 호출되도록 하였습니다.
```swift
    func showSuccessAlert(with juiceName: String) {
        let alertCountroll = UIAlertController(title: Phrases.noticeTitle.text, message: juiceName + Phrases.readyForJuice.text, preferredStyle: .alert)
        let okAction = UIAlertAction(title: Phrases.ok.text, style: .default, handler: nil )
        alertCountroll.addAction(okAction)
        present(alertCountroll, animated: false, completion: nil)
    }
    
    func showFailureAlert() {
        let alertCountrol = UIAlertController(title: Phrases.noticeTitle.text, message: Phrases.questionForStockChange.text, preferredStyle: .alert)
        let moveAction = UIAlertAction(title: Phrases.yes.text, style: .default, handler: { _ in self.moveStockChangeView() })
        let cancelAction = UIAlertAction(title: Phrases.no.text, style: .default, handler: nil )
        alertCountrol.addAction(moveAction)
        alertCountrol.addAction(cancelAction)
        present(alertCountrol, animated: false, completion: nil)
    }
```

### 화면전환을 모달을 사용할 것인가 네비게이션을 사용할 것인가?
처음 주신 프로젝트에서 네비게이션 컨트롤러가 이미 스토리보드에 나와있어, 그것을 사용하는 것이 맞다고 생각했습니다. 또한 네비게이션 컨트롤러를 사용할 경우 자동으로 뒤로가기 버튼이 만들어져 편리하다고 생각했습니다. 그러나, 우선 재고를 수정하는 화면 자체가 선택한 옵션에 대해서 집중해서 선택할 수 있는 화면에 더 가깝다고 생각이 되었고, HIG를 읽어보았을 때 모달이 이러한 특성을 잘 가지고 있다고 판단했습니다. 또한 아직 step2 구현 요구사항에 재고 수정 화면의 뒤로가기 버튼을 만들라는 이야기가 없었기 때문에 모달로 결정하여 진행하였습니다.

## 고민한 점 & 조언을 받고 싶은 부분
- extension 의 접근제어 
- id 값 처럼 사용할 수 있는 프로퍼티가 있는 것인지?
- 네비게이션안에서 모달이 가능한가? 
- FruitStore 네이밍에 관한 문제 
- 열거형 타입을 한 파일에 몰아 담는 방법에 대해 적절한 방법인가?

### extension 의 접근제어 
Juice 타입을 JuiceMake 에서 분리하면서, Juice 타입의 receip 연산프로퍼티를 extension으로 분리하였습니다. 이 과정에서 구조체 안에 extension을 넣을 수 없었습니다. 그래서 구조체 밖에 있는 extension에 의 receip 값을 JuiceMake 에서만 접근하도록 fileprivate 을 넣어주고 싶었습니다. 그러나 넣을 때와 안 넣을 때 모두 외부에서 접근이 가능했습니다. 에초에 enum extension에는 접근제어 구문을 넣을 수가 없는 것인지 궁금합니다. 

### id 값 처럼 사용할 수 있는 프로퍼티가 있는 것인지?
ios에서 주로 UIView(버튼 또는 라벨 등)을 구분할 때 주로 사용되는 키워드가 있는지 궁금합니다.(예를 들어, 안드로이드 view xml에서 id 같은 역할을 하는) 또한 ios의 tag의 값은 언제 주로 사용되는 값인지 궁금합니다.

### 네비게이션안에서 모달이 가능한가?
혹시 네이게이션 컨트롤러를 사용하면서, 모달의 특성을 가지고 창을 띄우는 것이 가능한가요? 

### FruitStore 네이밍에 관한 문제 
저희가 FruitStore 를 싱글톤 클래스로 사용하기 위해서 내부에 같은 이름을 가진 변수를 선언했습니다. 
`static var fruitStore = FruitStore()`
`JuiceMake` 나 `ViewCountroller` 에서 이 함수를 쉽게 가져다 쓰기 위해서 변수를 만들어 할당해 주었을 경우에, 저번에 이야기 해주셨듯이 `fruitStore` 라는 이름이 반복되는 패턴이 다시 일어났습니다. 
`private var fruitStore = FruitStore.fruitStore`
혹시 싱글톤 인스턴스의 이름을 정할 때의 조언을 구할 수 있을까요? 

### 열거형 타입을 한 파일에 몰아 담는 방법에 대해 적절한 방법인가?
현제 저희가 가지고 있는 열거형이 4개 입니다. 이 열거형들을 각각의 이름을 가진 파일을 만들어서 관리하는 것이 좋을까요? 아니면 모두 전역이고, 길지 않은 열거형이기 때문에 같은 파일에서 관리하여 주는 것이 좋을 까요?

cherrishRed committed:
feat : UILabel 객체 생성

Commit:
feat : getStock함수 추가

Stock의 값을 반환해주는 함수 추가

cherrishRed committed:
feat : FruitStore 클래스를 싱글톤으로 설정

Commit:
feat : fruitStore에서 데이터를 읽어와 UILabel에 표시

cherrishRed committed:
feat : 버튼에 태그 설정 및 orderButton 함수 생성

Commit:
feat : orderButton내부 로직 추가

tag를 받아 주스타입을 받아오는 기능 추가

cherrishRed committed:
feat : orderButton 메서드 내부에서 takeOrder 호출

Commit:
feat : 재고 업데이트 함수 updateFruitLable() 추가

cherrishRed committed:
feat : showAlert 함수 구현

Commit:
feat : showAlert()  상황에 따라 다른 얼럿문구가 표현되도록 수정

make함수 내부 로직 수정
takeOrder함수 리턴타입 추가 및 리턴값 추가
akeOrder함수의 리턴값을 담는 message변수 추가

cherrishRed committed:
feat : 재고에 따른 alert 표시 수정

Commit:
feat : NavigationController를 이용한 화면전환 기능 구현

cherrishRed committed:
refactor : moveStockCorrectionView 를 모달로 변경

Commit:
feat : 얼럿창에서 ‘예’ 선택시 재고수정화면으로 이동

cherrishRed committed:
refactor : IBOutlet 으로 버튼 값을 받도록 수정

이에 따른 switch 문 수정 및 분리, tag 사용 삭제

Commit:
feat : Juice타입 rawValue 값 설정

Commit:
Juice타입 rawValue 값 설정

Commit:
Merge remote-tracking branch 'origin/STEP2' into STEP2

cherrishRed committed:
feat : 에러 처리 함수 구현

Commit:
refactor : showAlert()함수 분리

showAlert()함수 showSuccessAlert()와 showFailureAlert()로 분리

cherrishRed committed:
refacfor : chekcError 함수 이름변경 및 위치 이동

Commit:
feat : 프로그램에 사용되는 문자열을 관리하는 Phrases타입 추가

공백제거
사용하지 않는 코드 삭제

cherrishRed committed:
refacfor : 전체적인 네이밍 변경

Commit:
feat : JuiceMaker에서 Juice타입 분리

JuiceMaker에선 extension이용하여 확장

cherrishRed committed:
feat : ViewController의 접근제어 수정

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,18 +7,95 @@
 
 class ViewController: UIViewController {
-
+    private var fruitStore = FruitStore.fruitStore
+    private var juiceMaker = JuiceMaker()
```
주스메이커 안에도 `fruitStore`이 있는데 여기서 또 선언해 주신 이유가 궁금해요!

cherrishRed replied:
저희가 싱글톤 패턴을 처음 써봐서, 어떻게 사용하는지에 대해 약간의 혼동이 있었던 것 같습니다. 네이밍이 길어보여서 접근시에 편하게 접근하려고 선언했었습니다. 
현재는 패턴 자체를 수정 하면서 저 부분은 없어 졌습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,18 +7,95 @@
+    private var fruitStore = FruitStore.fruitStore
+    private var juiceMaker = JuiceMaker()
+    
+    @IBOutlet weak var strawberryLabel: UILabel!
+    @IBOutlet weak var bananaLabel: UILabel!
+    @IBOutlet weak var pineappleLabel: UILabel!
+    @IBOutlet weak var kiwiLabel: UILabel!
+    @IBOutlet weak var magoLabel: UILabel!
+    
+    @IBOutlet weak var strawberryJuiceButton: UIButton!
+    @IBOutlet weak var bananaJuiceButton: UIButton!
+    @IBOutlet weak var pineappleJuiceButton: UIButton!
+    @IBOutlet weak var kiwiJuiceButton: UIButton!
+    @IBOutlet weak var mangoJuiceButton: UIButton!
+    @IBOutlet weak var strawberryAndBananaJuiceButton: UIButton!
+    @IBOutlet weak var mangoAndKiwiJuiceButton: UIButton!
```
혹시 왜 IBOutlet으로 선언하면 자동으로 느낌표가 붙을까요?

cherrishRed replied:
UILabel 이나 UIButton 의 구현부부 부분을 보니 필수로 초기에 설정해 주어야 하는 프로퍼티들과, 옵셔널 타입으로 된 프로퍼티들이 있더라구요. 옵셔널 타입이 프로퍼티의 값들의 할당이 되어있지 않을 수도 있어서 ! 를 사용해 바인딩을 해주는 것 같습니다. 
또한 ! 를 지우고 에러 구문을 확인해 보았을 떄, @IBOutlet 은 옵셔널 타입이 아니다. 라는 에러가 뜨는 걸로 봐서, @IBOutlet은 옵셔널값을 받을 수 없기에 강제 언래핑을 해 준것 같습니다.

stevenkim18 replied:
음.. IBOutlet은 옵셔널 타입이 아니라고 하셨는데 느낌표를 사용할 수 있다는 것은 옵셔널 타입이 가능한 걸로 알고 있는데 
제가 잘못 안 것 일까요?

혹시 몰라서 느낌표 대신 물음표를 붙여봤는데도 잘 컴파일이 되었습니다.

--- 

제가 생각해본 바로는 보통 느낌표를 사용하는 경우는 개발자가 확실하게 여기에는 값을 집어 넣을 것이다 라고 판단 했을 때 입니다.

보통 iboutlet을 사용하면 반드시 스토리보드에서 뷰를 연결해야 컴파일이 됩니다.

뷰와 iboutlet 잘 연결 되어 있지 않으면 앱 시작 바로 애러가 납니다. 그래서 스토리 보드에 뷰를 잘 연결 했다는 가정 하에 iboutlet를 사용하기 위함이 아닐까 싶습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,18 +7,95 @@
+        updateFruitLable()
+    }
+    
+    private func matchButtonToJuice(_ button: UIButton) -> Juice? {
```
함수 이름과 매개변수에 button이 두번 들어갔네요 어떻게 수정하면 좋을까요?

cherrishRed replied:
네, 이름을 `matchJuice(with button: UIButton)` 으로 수정하였습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,18 +7,95 @@
+    }
+    
+    private func matchButtonToJuice(_ button: UIButton) -> Juice? {
+        var juice: Juice?
+        switch button {
+        case strawberryJuiceButton:
+            juice = .strawberryJuice
+        case bananaJuiceButton:
+            juice = .bananaJuice
+        case pineappleJuiceButton:
+            juice = .pineappleJuice
+        case kiwiJuiceButton:
+            juice = .kiwiJuice
+        case mangoJuiceButton:
+            juice = .mangoJuice
+        case strawberryAndBananaJuiceButton:
+            juice = .strawberryAndBananaJuice
+        case mangoAndKiwiJuiceButton:
+            juice = .mangoAndKiwiJuice
+        default:
+            juice = nil
+        }
+        return juice
+    }
```
juice라는 지역변수를 두지 않고 case에서 바로 return 해도 되지 않을까요?

cherrishRed replied:
그게 더 간결해 보일 거 같네요. 수정 했습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,18 +7,95 @@
-        juiceMaker.takeOrder(.mangoJuice)
-        juiceMaker.takeOrder(.mangoJuice)
-        juiceMaker.takeOrder(.mangoJuice)
+        // Do any additional setup after loading the view.
```
주석이 꼭 필요한가요?

cherrishRed replied:
자동으로 생성되는 주석이라 지우지 않았는데...신경을 못 쓴것 같습니다. 삭제했습니다!

stevenkim18 replied:
주석은 중간 커밋에는 들어가도 좋으나 최종 PR에서는 신경 써 주시면 감사하겠습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -7,9 +7,10 @@
 import Foundation
 
 class FruitStore {
+    static var fruitStore = FruitStore()
     private var stock: [Fruit: Int]
-
```
이번 PR에서 가능 궁금한 부분은 왜 fruitStore 싱글톤으로 선언하셨는지 입니다.
- 싱글톤은 언제 사용될까요?
- 장단점은 무엇인가요?
- Swift에서는 싱글톤으로 구현된 객체는 무엇이 있나요??

cherrishRed replied:
**싱글톤은 언제 사용될까요?**
싱글톤은 여러곳에서 같은 인스턴스에 접근해 모두가 같이 값을 변경해야 하는 경우 사용됩니다. 

**장단점은 무엇인가요?**
장점은 어디에서나 접근할 수 있는 전역인스턴스 이기 때문에 코드가 간결하고 편합니다.  또한 메모리를 절약할 수 있습니다. 
단점은 객체지향 설계 원칙에 어긋납니다. 객체지향의 원칙 중 개방 폐쇠 원칙으로 MVC 각각의 역할이 잘 분리되어야 하는데, 싱글톤을 사용할 경우 이 셋의 결합도가 올라가 각각의 역할이 모호해 지게 됩니다. 또한 공부했을 때 멀티 쓰레드 환경에서 동기화 처리를 하지 않을 경우 2개의 인스턴스가 생성되는 문제점이 있다고 들었습니다. 

**Swift에서는 싱글톤으로 구현된 객체는 무엇이 있나요??**
노티피케이션 센터, NSFileManager, NSWorkspace, UIApplication 그리고 UIAccelerometer 등이 있다고 조사를 했습니다. 
특히 이번에 노티피케이션 센터를 적용해서 사용해 보면서, 싱글톤 패턴이 사용되었다는 것을 체감할 수 있었습니다.

stevenkim18 replied:
조금 번거로운 일이지만  새로운 기술을 사용할 때는 간단히
- 왜 사용하는지
- 사용 예시
- 장단점
등은 공부하고 가시면 좋아요!

조사 하셨던것 처럼 뭔가 일을 해주는 것들이 Swift에서는 싱글톤으로 구현되어있어요. 
재고를 관리하는 것은 데이터를 저장하는 것인데 이를 싱글톤으로 해주것은 적절하지 않아 보입니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,22 @@
+
+import Foundation
+
+enum Juice: String {
+    case strawberryJuice = "딸기쥬스"
+    case bananaJuice = "바나나쥬스"
+    case pineappleJuice = "파인애플쥬스"
+    case kiwiJuice = "키위쥬스"
+    case mangoJuice = "망고쥬스"
+    case strawberryAndBananaJuice = "딸바쥬스"
+    case mangoAndKiwiJuice = "망키쥬스"
+    
+    var koreanName: String {
+        return self.rawValue
+    }
+}
```
`CustomStringConvertible` 을 사용하면 코드를 더 줄일 수 있어요!! 한번 학습해보세요!

Siwon-L replied:
CustomStringConvertible를 찾아보고 적용해 보았는데, 코드가 줄어드는 느낌은 받지 못하였습니다. 좀 더 CustomStringConvertible를 잘 활용할 수 있는 방법을 알고 싶습니다.

stevenkim18 replied:
`"\(Juice.starberryJuice)"` 처럼 열거형에 case를 문자열로 바뀌면 지정해 놓은 단어가 출력되지 않나요? "딸기쥬스" 처럼요!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,44 +6,49 @@
 
 import Foundation
 
+enum JuiceMakeError: Error {
+    case lackOfStock
+}
```
PR에서도 언급했지만 케이스가 한개인데 애러 타입을 정의하신 이유가 궁금합니다!

Siwon-L replied:
지금은 한 개이지만, 스탭 3에서 오류가 추가될 수도 있다는 생각에 미리 좀 구현해두었던 부분이었는데 레드와 이야기해 본 결과, 추가적인 에러가 생기지 않은 것 같고, 만약 추가적인 에러가 생긴다면 그때 다시 에러 핸들링을 하는 방향으로 얘기를 나눴고, 때문에 Bool 타입으로 처리를 해주고 있습니다.

cherrishRed replied:
사실은 에러핸들링을 연습해 보고 싶어서 이 부분을 적용했는데요, 상의 끝에 에러 핸들링이 필요없다고 생각되어 삭제 했습니다.

stevenkim18 replied:
확장성을 고려해서 Error 타입을 사용하셨다면 이유가 타당한 것 같습니다.

이 부분도 언제 Error 타입을 쓰면 좋을지 아니면 그냥 Bool 리턴함수로 처리하면 좋을지도 많이 고민해보세요!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,44 +6,49 @@
+    case lackOfStock
+}
+
+extension Juice {
+    var recipe: [(Fruit, Int)] {
+        switch self {
+        case .strawberryJuice:
+            return [(.strawberry, 16)]
+        case .bananaJuice:
+            return [(.banana, 2)]
+        case .pineappleJuice:
+            return [(.pineapple, 2)]
+        case .kiwiJuice:
+            return [(.kiwi, 3)]
+        case .mangoJuice:
+            return [(.mango, 3)]
+        case .strawberryAndBananaJuice:
+            return [(.strawberry, 10), (.banana, 1)]
+        case .mangoAndKiwiJuice:
+            return [(.mango, 2), (.kiwi, 1)]
+            
+        }
+    }
+}
```
주스 메이커 파일인데 왜 주스 extension이 여기 있나요?

Siwon-L replied:
Juice 안에 주스 이름이 있고 주스의 이름은 주스 메이커에서도 사용되지만, 뷰 컨트롤에서도 사용되어 따로 전역으로 빼주는 대신 주스 메이커 안에서만 사용되는 recipe에 대한 정보는 Juice를 주스 메이커 내부에서 확장하여 사용했습니다. 주스 메이커에서만 사용되는 recipe에 대한 데이터를 전역으로 빼는 것보다 주스 메이커 내부에 있는 것이 더 타입의 역할에 부합하다고 생각했습니다.

stevenkim18 replied:
그럼 만약 Swift 파일이 엄청 많아졌다고 가정을 해보겠습니다.

주스의 종류가 다양해지면 Juice 파일에서도 수정하고 JuiceMaker 파일에 가서도 수정을 해야 합니다. 

한 파일에 있을 때와 아닐 때의 어떤 것이 유지보수가 더 편할까요?

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,44 +6,49 @@
-                return [(.mango, 2), (.kiwi, 1)]
-                
-            }
+    func orderAndCheck(_ juice: Juice) -> Bool {
+        do {
+            try takeOrder(juice)
+        } catch {
+            return false
         }
+        return true
     }
```
보통 함수는 한가지 일만 하는 것이 좋습니다.

지금 네이밍을 보면 and 가 있어서 두가지 일을 하는데 어떻게 수정하면 좋을까요?

Siwon-L replied:
오류 처리를 제거하면서 위 함수는 삭제되었고, 위 기능을 canMake 함수가 대신하게 되었습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,44 +6,49 @@
+            throw JuiceMakeError.lackOfStock
         }
     }
     
```
```swift
private func ensureStock(of juice: Juice) -> Bool {
        var checkList: [Bool] = []
        for ingredient in juice.recipe {
            checkList.append(fruitStore.isEnoughStock(of: ingredient))
        }
        return checkList.allSatisfy{ $0 }
    }
```

`ensureStock` 이라는 함수 이름도 한번 고민해 봅시다!
bool 타입을 리턴하고 있는데요.
아래 링크를 참고해서 더 좋은 네이밍을 고민해주세요!
https://soojin.ro/blog/english-for-developers-swift

(ensureStock 부분이 태그가 안되서 바로 위줄에 커멘트를 달았습니다!)

Siwon-L replied:
링크를 참고해서 고민해 본 결과 canMake라는 함수명이 적합하다고 생각했습니다. "만들 수 있다."라는 기능과 명확하게 떨어지는 부분 또한 있었습니다.

stevenkim18 replied:
꼭 위 링크대로 만드시라는 의미는 아니였습니다!! 저 좋은 네이밍을 많이 고민해 보시라는 차원에서 첨부했습니다!

stevenkim18 reviewed (commented):
사파리 레드!! 이번에도 수고 많으셨습니다!!

대체적으로 잘 구현하신 것 같아요!! 

아 그리고 제출해 주실때 main 브랜치로 머지 한 다음에 PR 날려주시면 감사할 거 같아요. 

main 브랜치에 있으면 클론 하기가 편합니다! ㅎㅎ

### UML과 문서에 대해서..

- UML은 너무 잘 봤습니다!! 한눈에 이해하기가 편했습니다.
    - 만약 UML 그리는 시간이 오래 걸리신다면 간단하게만 그려주셔도 좋고 손으로 그린 것을 첨부해주셔도 좋습니다. 너무 예쁘게 안해주셔도 됩니다 ㅎㅎ
- view 부분도 UML로 주셨는데 이 부분은 구현한 화면을 캡처해주시거나 동작을 하는 gif까지 첨부하셨으면 더 좋았을 것 같아요!! 그러면 제가 딱 PR을 봤을 때 한눈에 라벨과 버튼들이 이렇게 사용되었구나 라고 쉽게 알 수 있었을 것 같아요!! 계속 어떻게 효율적으로 잘 보여줄지를 고민해주세요!!

### ****어려 버튼에 대한 액션을 하나의 IBAction으로 처리해줄 것인가****

- 한개의 함수로 처리하시다는 생각은 너무 잘하셨습니다.
- 그리고 tag와 버튼의 title을 사용하는 방법을 비교하신 것도 너무 잘하셨습니다.
- 분명 이 방법만은 있지는 않을 것 같아요!! 어떻게 다른 방법으로 구현할 수 있는지도 찾아보시고 다른 팀하고도 많이 대화나눠보세요!!

### ****어떤 방식으로 에러 처리를 해줄 것인가****

- Error 타입에 case가 한개 던데 단지 “오류가 발생하는 것을 보여주고 싶어서.." 라는 이유만 있는건지 case가 추가되는 것을 고려해서 만드신 것인지가 궁금합니다.
    - 지금 상황에서는 if else로도 충분히 구현할 수 있을 것 같습니다. 오히려 do - catch 을 사용해서 뎁스가 생겨버셨네요.
    - 그리고 catch 부분에서도 그냥 return 을 해서 애러 타입을 정의해준 의미가 크게 없어 보입니다!!
    - 어떻게 생각하시는지 궁금합니다!!

### ****다른 상황에 보이는 얼럿에 대한 함수를 분리할 것인가? 통합할 것인가?****

- 일단 먼저 구현하신 것은 잘하셨습니다!! 이렇게 먼저 구현을 해보시고 고민해보시는 것은 좋습니다!
- 저 같은 경우는 일단 매개변수를 여러개 사용해서 title message alert action을 추가해줄 수 있을 것 같아요! 이부분은 많이 해보시면 경험이 쌓여서 한개의 함수로도 처리할 수 있을 것이라고 생각됩니다!
- 아니면 더 여러개의 함수를 쪼개어서 구현할 수도 있을 것 같아요!

### ****화면전환을 모달을 사용할 것인가 네비게이션을 사용할 것인가?****

- 둘다 가능합니다!! 이러저리 다양하게 구현해보시고 차이점을 공부해보세요!

### ****extension 의 접근제어****

- 이부분은 코드를 같이 첨부해주시면 더 이해가 잘 될 것 같아요!! 안되었던 코드랑 되는 코드를 알려주세요!!

### ****id 값 처럼 사용할 수 있는 프로퍼티가 있는 것인지?****

- 안드로이드 처럼 구현 할 수 있는 방법은 Ios에서는 없습니다.
- 물론 각 뷰가 고유의 id를 내부적으로는 가지고 있을거에요. 그런데 스토리보드를 사용하거나 직접 코드로 인스턴스를 만들어야 사용가능합니다.
- tag는 저는 잘 안써봤는데 뷰를 구분할 때 사용할 수 있는 방법 중에 하나 입니다.

### ****네비게이션안에서 모달이 가능한가?****

- 지금 주어진 스토리 보드 예시는 다른 각각 다른 네비게이션 컨트롤러 안에 있습니다.
- 네비게이션 컨트롤러안에서 다른 네비게이션 컨트롤러를 가지고 있는 뷰컨을 모달로 띄우는 것은 가능합니다. 하지만 모달에서 뒤로가기를 눌렀을 때는 전체 네비게이션 컨트롤러를 가지고 있는 뷰컨이 사라지는 것입니다.
- 모달에 대해서 조금 더 공부해보세요!

### ****FruitStore 네이밍에 관한 문제****

- Swift 내부에서 구현되어 있는 싱글톤 객체는 주로 `shared` 라는 네이밍을 많이 사용합니다.
- 그런데 위와 같은 상황이면 중복을 피하기 위해서 stock 과 같은 단어를 사용할 거 같아요!

### ****열거형 타입을 한 파일에 몰아 담는 방법에 대해 적절한 방법인가?****

- 때에 따라 다른 것 같습니다.
- 저라면 4개의 파일이면 enum이라는 폴더를 만들고 하위에 둘 것 같습니다.

Commit:
refactor : FruitStore타입의 싱글톤 인스턴스 네이밍 수정

싱글톤 인스턴스 네이밍 fruitStore -> shared
JuiceMaker과 ViewController의 싱글톤 변수 삭제 및 공백 제거

Commit:
Revert "refactor : FruitStore타입의 싱글톤 인스턴스 네이밍 수정"

This reverts commit 2bd23415301b5e94679cc10f4dbe550ee3a0d486.

Commit:
Merge branch 'STEP2' of https://github.com/saafaaari/ios-juice-maker into STEP2

Commit:
refactor : FruitStore타입의 싱글톤 인스턴스 네이밍 수정

싱글톤 인스턴스 네이밍  fruitStore -> shared 변경
JuiceMaker와 ViewController의 싱글톤 변수 삭제

cherrishRed committed:
refactor : matchButtonToJuice 메서드 안의 불필요한 변수 삭제

추가로 불필요한 주석 제거

Commit:
feat : 싱글톤 패턴에서 KVO로 수정

cherrishRed committed:
refactor : KVO 삭제, NotificationCenter 추가

Commit:
feat : FruitStore타입 내부  NotificationCenter.default.post 추가

들려쓰기 정렬

cherrishRed committed:
refactor : updateFruitLable 로직 변경

Commit:
feat : 라벨을 초기화 해주는 함수 initLable() 추가

주석 제거
Fruit타입 CaseIterable 프로토콜 채택
컨벤션 수정 및 FruitStore타입 NSObject 제거
싱글톤 패턴 제거

cherrishRed committed:
refactor : ensureStock 네이밍 canMake 으로 변경

Commit:
refactor : matchButtonToJuice네이밍 변경

matchButtonToJuice -> matchJuice 변경 및 with전달 인자 추가

cherrishRed committed:
refactor : 에러 핸들링 삭제 및 불필요한 함수 제거

Commit:
refactor : Juice타입 CustomStringConvertible프로토콜 채용

Commit:
Revert "refactor : Juice타입 CustomStringConvertible프로토콜 채용"

This reverts commit ffc6ff25bda02a9cd0cfae26fc4ac5efa57c506e.

Commit:
Merge branch 'STEP2' of https://github.com/saafaaari/ios-juice-maker into STEP2

Commit:
refactor : Juice타입 CustomStringConvertible프로토콜 채용

cherrishRed commented:
@stevenkim18

안녕하세요 steven 😁 🤚🏻

STEP2 리펙토링 완료 했습니다. 

### 변경한 부분

- 싱글톤을 제거 하고, 노티피케이션 센터를 이용하여 값 전달
- 에러핸들링 제거

### 조언을 얻고 싶은 부분

- KVO 가 사용자 지정 타입에서 작동하지 않는 이유가 있나요?
- notification 으로 일어난 값초기화 부분을 더욱 간결히 처리할 방법이 있을까요?
- 추천해주신 customStringConvertible 을 사용하였는데 코드가 줄어드는 것을 체감적으로 느끼진 못했습니다. 지금 저희 코드가 길지 않아서 그런 걸까요?

### KVO 가 사용자 지정 타입에서 작동하지 않는 이유가 있나요?

노티피케이션 센터를 차용하기 전에, FruitStore 의 stock 프로퍼티만 사용하면 되었기 때문에 KVO 패턴을 먼저 도전했었습니다. 

그런데, 

```swift
class FruitStore: NSObject {
    @objc dynamic var stock: [Fruit: Int]
}
```

Fruit 이라는 사용자 지정 enum 타입이 Object-C 를 만족할 수 없다는 오류가 생기더라구요.  NSObject를 상속할수 없어서 일어나는 문제 인걸로 확인했습니다. 이 경우에 KVO를 사용할 수 없는 것일까요?

### notification 으로 일어난 값초기화 부분을 더욱 간결히 처리할 방법이 있을까요?

viewController에 있는 Label의 값을 노티피케이션을 받아 업데이트 하는 함수

```swift
@objc private func updateFruitLable(_ notification: Notification) {
        guard let fruitName = notification.userInfo?["fruit"] as? Fruit else { return }
        guard let fruitStock = notification.userInfo?["stock"] as? Int else { return }
        if fruitName == .strawberry {
            strawberryLabel.text = String(fruitStock)
        } else if fruitName == .banana {
            bananaLabel.text = String(fruitStock)
        } else if fruitName == .mango {
            magoLabel.text = String(fruitStock)
        } else if fruitName == .kiwi {
            kiwiLabel.text = String(fruitStock)
        } else if fruitName == .pineapple {
            pineappleLabel.text = String(fruitStock)
        }
    }
```

JuiceMake 에 있는 initializeLable()함수로 라벨 초기값을 위한 함수

```swift
func initializeLable() {
        for fruit in Fruit.allCases {
            fruitStore.changeStock(of: fruit, to: .zero)
        }
    }
```

FruitStore 에 있는 changeStock()함수로 스톡이 변화할 때 마다 post를 보내주는 함수
```swift
func changeStock(of fruit: Fruit, to numberOfFruit: Int) {
        guard var inStock: Int = stock[fruit] else { return }
        inStock += numberOfFruit
        stock[fruit] = inStock
        NotificationCenter.default.post(name: Notification.Name("notificationStock"),
                                        object: nil,
                                        userInfo: ["fruit": fruit,
                                                   "stock": inStock])
    }
```

stevenkim18 reviewed (commented):
수정하신 부분 잘 봤습니다!!

NotificationCenter를 사용하신 시도는 정말 잘하셨습니다!!질

그런데 궁금한 부분은 `FruitStore` 가 참조 타입이라서 한개의 인스턴스만을 생성하면 그 부분안에서 재고 관리를 할 수 있을 것 같습니다.

NotificationCenter 장점도 있지만 단점이 명확합니다.

1. 디버깅하기가 힘듭니다.. 어디서 보내서 어디서 받는지 나중에 알기가 어렵습니다.
2. 질문 주신 것 처럼 NotificationCenter가 값을 주고 받는 것이 딕셔너리의 value가 Any 타입이기 때문에 무조건 형변환을 해줘야합니다. 그리고 post를 보낼 때도 일일히 다 보내줘야 합니다. (물론 Notification.name이나 userInfo를 변수를 빼서 관리하면 조금 더 코드는 깔끔해 질 것 입니다)

그래서 이번 프로젝트에서는 제 개인적인 생각에는 적절해 보이지 않습니다!

꼭 `FruitStore` 로 관리할 수 있을 지를 고민해 봐주시고 적용도 해보시면 좋겠습니다!!

(만약 너무 어려우시면 꼭 디코 음성채널에 초대해주세요!!)

---

그리고 KVO는 저도 사용은 잘 안해봤지만.. 아마 Friut가 열거형이라서 그런 것 같습니다. 열거형도 값타입입니다. KVO는 주로 참조타입에서만 사용가능한 걸로 알고 있습니다.

 

마지막으로 **`customStringConvertible`** 

ViewController 29줄에서

```swift
showSuccessAlert(with: String(describing: juice))
```

를

```swift
showSuccessAlert(with: "\(juice)")
```

로 바뀌줄 수 있습니다!

**월요일에 한번 간단히 디코 하시고 그 때 위 부분들에 대해서 이야기 나뉘보고 머지해드리겠습니다!!**
